core/src/main/java/org/elasticsearch/index/similarity/ScriptedSimilarity.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.similarity;
+
+import org.apache.lucene.index.FieldInvertState;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.search.CollectionStatistics;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.TermStatistics;
+import org.apache.lucene.search.similarities.Similarity;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.SmallFloat;
+import org.elasticsearch.script.ExecutableScript;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+/**
+ * A {@link Similarity} implementation that allows scores to be scripted.
+ */
+public final class ScriptedSimilarity extends Similarity {
+
+    private final String scriptString;
+    private final Supplier<ExecutableScript> scriptSupplier;
```

Comment:
We should be using a new script context here. Then this can be SimilarityScript.Factory.  The new context can return `float` directly, and take Stats as an arg.